File: cargo/bindings/swift/ExploreViewData.swift
//
//  ExploreViewData.swift
//  BuoyFinder
//
//  Created by Matthew Iannucci on 8/28/18.
//  Copyright © 2018 Matthew Iannucci. All rights reserved.
//

import Foundation

class ExploreViewData {
    
    private let raw: OpaquePointer
    
    init() {
        raw = explore_view_data_new()
    }
    
    deinit {
        explore_view_data_free(raw)
    }
    
    var stationCount: Int {
        get {
            return Int(explore_view_data_station_count(raw))
        }
    }
    
    func station(at index: Int) -> BuoyStation {
        let rawStation = explore_view_data_station_index(raw, Int64(index))
        return BuoyStation(rawStation!)
    }
    
    var stations: [BuoyStation] {
        get {
            var stations: [BuoyStation] = []
            for i in 0..<stationCount {
                let buoy = station(at: i)
                stations.append(buoy)
            }
            return stations
        }
    }
}

File: cargo/bindings/java/com/mpiannucci/buoyfinder/core/ExploreViewData.kt
package com.mpiannucci.buoyfinder.core

class ExploreViewData {

    val ptr: Long

    constructor() {
        ptr = new()
    }

    constructor(rawPtr: Long) {
        ptr = rawPtr
    }

    fun finalize() {
        free(ptr)
    }

    val stationCount: Long
        get() = stationCount(ptr)

    fun station(index: Long): BuoyStation {
        val rawStation = stationAtIndex(ptr, index)
        return BuoyStation(rawStation)
    }

    private external fun new(): Long
    private external fun free(ptr: Long)
    private external fun stationCount(ptr: Long): Long
    private external fun stationAtIndex(ptr: Long, index: Long): Long
}